README.md
# Date With Offset

In JavaScript, all `Date`s have a local time zone. On my computer:

```javascript
var now = new Date();
// Sun Apr 14 2013 09:49:16 GMT-0700 (PDT)
```

This makes working with time zones difficult. You can represent that date in
UTC with `Date.prototype.toISOString`:

```javascript
now.toISOString();
// "2013-04-14T16:49:16.576Z"
```

Unfortunately, you can't pass around an actual `Date` in any other time zone.
Instead, create a `DateWithOffset`:

```javascript
var nowInUTC = new DateWithOffset(0);
// Sun Apr 14 2013 16:49:16 GMT+0000
```

## Creating `DateWithOffset`s

The `DateWithOffset` constructor works just like the `Date` constructor, but
the *last* argument is always the offset from UTC in minutes. Some examples:

```javascript
var nowInParis = new DateWithOffset(60);
// Sun Apr 14 2013 17:49:16 GMT+0100

var theSameTimeInMelbourne = new DateWithOffset(nowInParis, 600);
// Mon Apr 15 2013 02:49:16 GMT+1000
```

### Date Parsing

If the first argument is a String and contains an offset end with "Z",
it is treated as UTC time:

```
var newYearsGMTInBoston = new DateWithOffset("Jan 1 2013 00:00Z", -300);
// Mon Dec 31 2012 19:00:00 GMT-0500
```

If it's a String and doesn't contain an offset of end with "Z", it is treated
as local to the given offset:

```
var newYearsInBoston = new DateWithOffset("Jan 1 2013 00:00", -300);
// Tue Jan 01 2013 00:00:00 GMT-0500
```

Similarly, `DateWithOffset`s created with individual year, month, and day
(and, optionally, hours, minutes, seconds, and milliseconds) arguments are
treated as local to the given offset:

```
var newYearsInChicago = new DateWithOffset(2013, 0, 1, -360);
```

**Note** this behavior differs from that of the normal `Date` constructor,
which treats such strings as local to the *browser* (or server execution
environment).

### Daylight Savings Time

Many time zones obey Daylight Savings. This means that they can't be
repesented as a single number; rather, they are a `function` that returns
a different offset based on the `Date` passed in.

Imagine a mythical time zone that is three hours ahead (East) of GMT from
November through March each year, and four hours ahead April through October.
If you represent this time zone with the following function:

```javascript
function mythicalTimeZone(date) {
  var month = date.getMonth();
  return (month < 3 || month > 9) ? 180 : 240;
}
```

you could create dates in this time zone as follows:

```javascript
var date = new DateWithOffset(2013, 5, 16, mythicalTimeZone); // June 16 2013, MDT
date.getTimezoneOffset(); // -240
```

If you *modify* the date, it will use the correct offset for the new time:

```javascript
date.setMonth(1); // February 16, 2013, MST
date.getTimezoneOFfset(); // -180
```

### Rich Offset Objects

The last argument can be a `Number` or `function` (as above) or anything that
responds to `valueOf`. If you have richer time zone objects, you can pass them
directly into `new DateWithOffset`:

```javascript
var tokyo = {
  name: 'Tokyo',
  toString: function() { return 'Tokyo (GMT+0900)' },
  valueOf: function() { return 540; }
};

var nowInTokyo = new DateWithOffset(now, tokyo);
// Mon Apr 15 2013 01:49:16 GMT+0900
```

***Note***: the offset is that between *this* object and *UTC*, which means
that it is positive if the object's time zone is ahead of UTC and negative
if it is behind. This is the opposite of what
[`Date.prototype.getTimezoneOffset`](https://developer.mozilla.org/en-US/docs/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset)
returns.

## Compatibility with `Date`

You can use a `DateWithOffset` anywhere you use a `Date`:

```javascript
nowInUTC.getHours();                  // 16
nowInParis.getTime();                 // 1365958156000
theSameTimeInMelbourne.getTime();     // 1365958156000
newYearsInBoston.getTimezoneOffset(); // 300

newYearsInBoston.setDate(15);
newYearsInBoston;                     // Tue Jan 15 2013 00:00:00 GMT-0500
```

## Additional Features

Get back the original offset:

```
nowInBoston.offset();
// -300

nowInTokyo.offset().toString();
// "Tokyo (GMT+0900)"
```

Get a new `DateWithOffset` representing the same point in time at a
different UTC offset:

```
var nowInChicago = nowInBoston.withOffset(-300)
```

Get a plain `Date` representing the same point in time at the local offset:

```
var nowInLocal = nowInParis.date();
```

## Related Projects

[node-time](https://github.com/TooTallNate/node-time) provides very similar
functionality with a different API. It supports time zone names (not just
offsets), but only runs in Node.

If you want time zone parsing support, try
[timezone-js](https://github.com/mde/timezone-js) or
[timezone](https://npmjs.org/package/timezone).

If you want a richer library for parsing, validating, manipulating, and
formatting dates, try [Moment.js](http://momentjs.com/).

If all you need to do is map Rails time zone names to IANA ones, you'll love
[rails-timezone-js](https://github.com/davidwood/rails-timezone-js).


Changelog.md
## v1.1.0, 2013-09-07

 * add `DateWithOffset.prototype.toJSON`

## v1.0.1, 2013-04-15

 * fix `module` / `window` existence checks
 * add "related projects" list to README

## v1.0.0, 2013-04-14

Initial release

test/offset_test.js
var assert = require("assert"),
    DateWithOffset = require('../');

function assertSameInstant(actual, expected) {
  "use strict";

  actual   = new Date(actual);
  expected = new Date(expected);

  assert.equal(
    +actual,
    +expected,
    "expected " + actual.toISOString() +
    " to be the same time as " + expected.toISOString()
  );
}

describe('DateWithOffset', function() {

  "use strict";

  var halfNoonUTC = new Date("2008-12-31T12:30Z"),
      chatham = {
        toString: function() { return 'Chatham (+12:45)'; },
        valueOf: function() { return 765; }
      },
      inChatham;

  beforeEach(function() {
    inChatham = new DateWithOffset(halfNoonUTC, chatham);
  });

  describe('#toString', function() {
    it('returns a Date string in the correct offset', function() {
      var expected = 'Thu Jan 01 2009 01:15:00 GMT+1245';
      assert.equal(inChatham.toString(), expected);
    });
  });

  describe('#toJSON', function() {
    it('returns an ISO string in UTC', function() {
      var expected = halfNoonUTC.toJSON();
      assert.equal(inChatham.toJSON(), expected);
    });
  });

  describe('#valueOf', function() {
    it('returns the same moment in time', function() {
      assertSameInstant(inChatham, halfNoonUTC);
    });
  });

  describe('#offset', function() {
    it('returns the original offset', function() {
      assert.strictEqual(inChatham.offset(), chatham);
    });
  });

  describe('#getTimezoneOffset', function() {
    it('returns the inverse of the value of the supplied offset', function() {
      assert.equal(inChatham.getTimezoneOffset(), -765);
    });
  });

  describe('#withOffset', function() {
    it('returns a new DateWithOffset representing the same time with the given offset', function() {
      var inParis = inChatham.withOffset(60);
      assert.ok(inParis instanceof DateWithOffset, 'expected ' + inParis + ' to be a DateWithOffset');
      assertSameInstant(inParis, inChatham);
      assert.equal(inParis.offset(), 60);
      assert.equal(inParis.getHours(), 13);
    });
  });

  function assertGetter(property, expectedValue) {
    describe('#get' + property, function() {
      it('returns the correct value', function() {
        var actual = inChatham['get' + property]();
        assert.equal(actual, expectedValue);
      });
    });
  }

  assertGetter('Date',          1);
  assertGetter('Day',           4);
  assertGetter('FullYear',      2009);
  assertGetter('Hours',         1);
  assertGetter('Milliseconds',  0);
  assertGetter('Minutes',       15);
  assertGetter('Month',         0);
  assertGetter('Seconds',       0);
  assertGetter('Year',          109);

  assertGetter('Time', halfNoonUTC.getTime());

  assertGetter('UTCDate',         halfNoonUTC.getUTCDate());
  assertGetter('UTCDay',          halfNoonUTC.getUTCDay());
  assertGetter('UTCFullYear',     halfNoonUTC.getUTCFullYear());
  assertGetter('UTCHours',        halfNoonUTC.getUTCHours());
  assertGetter('UTCMilliseconds', halfNoonUTC.getUTCMilliseconds());
  assertGetter('UTCMinutes',      halfNoonUTC.getUTCMinutes());
  assertGetter('UTCMonth',        halfNoonUTC.getUTCMonth());
  assertGetter('UTCSeconds',      halfNoonUTC.getUTCSeconds());

  function assertSetter(property, newValue, expectedNewISOString) {
    describe('#set' + property, function() {
      it('updates the date', function() {
        inChatham['set' + property](newValue);
        assertSameInstant(inChatham, expectedNewISOString);
      });
    });
  }

  assertSetter('Date',         4,    '2009-01-04T01:15:00+1245');
  assertSetter('FullYear',     2010, '2010-01-01T01:15:00+1245');
  assertSetter('Hours',        17,   '2009-01-01T17:15:00+1245');
  assertSetter('Milliseconds', 230,  '2009-01-01T01:15:00.230+1245');
  assertSetter('Minutes',      44,   '2009-01-01T01:44:00+1245');
  assertSetter('Month',        7,    '2009-08-01T01:15:00+1245');
  assertSetter('Seconds',      19,   '2009-01-01T01:15:19+1245');
  assertSetter('Year',         102,  '2002-01-01T01:15:00+1245');

  assertSetter('Time', (new Date('2017-03-15T11:52:18+1245')).getTime(), '2017-03-15T11:52:18+1245');

  assertSetter('UTCDate',         4,    '2008-12-04T12:30:00.000Z');
  assertSetter('UTCFullYear',     2010, '2010-12-31T12:30:00.000Z');
  assertSetter('UTCHours',        17,   '2008-12-31T17:30:00.000Z');
  assertSetter('UTCMilliseconds', 230,  '2008-12-31T12:30:00.230Z');
  assertSetter('UTCMinutes',      44,   '2008-12-31T12:44:00.000Z');
  assertSetter('UTCMonth',        7,    '2008-08-31T12:30:00.000Z');
  assertSetter('UTCSeconds',      19,   '2008-12-31T12:30:19.000Z');

});


test/constructor_test.js
var assert = require("assert"),
    DateWithOffset = require('../');

describe('new DateWithOffset', function() {

  "use strict";

  it('requires at least one argument', function() {
    assert.throws(
      function() {
        new DateWithOffset();
      },
      TypeError
    );
  });

  it('requires an offset that can be converted to a Number', function() {
    assert.throws(
      function() {
        new DateWithOffset('foo');
      },
      TypeError
    );
  });

  it('accepts a number offset', function() {
    assert.doesNotThrow(function() {
      new DateWithOffset(-400);
    });
  });

  it('accepts an offset that responds to valueOf', function() {
    var offset = { valueOf: function() { return 240; } };
    assert.doesNotThrow(function() {
      new DateWithOffset(offset);
    });
  });

  describe('with a Date and an offset', function() {
    it('returns the same moment in time in the given offset', function() {
      var date    = new Date("2008-11-22T12:00:00Z"),
          inTokyo = new DateWithOffset(date, 540);
      assert.equal(inTokyo.getTime(), date.getTime());
      assert.equal(inTokyo.getHours(), 21);
    });
  });

  describe('with a timestamp and an offset', function() {
    it('returns the same moment in time in the given offset', function() {
      var timestamp = (new Date("2008-11-22T12:00:00Z")).getTime(),
          inTokyo   = new DateWithOffset(timestamp, 540);
      assert.equal(inTokyo.getTime(), timestamp);
      assert.equal(inTokyo.getHours(), 21);
    });
  });

  describe('with a GMT date string and an offset', function() {
    it('returns the same moment in time in the given offset', function() {
      var isoString = "2008-11-22T12:00:00.000Z",
          inTokyo   = new DateWithOffset(isoString, 540);
      assert.equal(inTokyo.toISOString(), isoString);
      assert.equal(inTokyo.getHours(), 21);
    });
  });

  describe('with a local date string and an offset', function() {
    it('treats the time as local to the given offset', function() {
      var inTokyo = new DateWithOffset('Nov 22 2008 21:00:00', 540);
      assert.equal(inTokyo.toISOString(), '2008-11-22T12:00:00.000Z');
      assert.equal(inTokyo.getHours(), 21);
    });
  });

  describe('with year, month, day, hours, minutes, and an offset', function() {
    it('treats the time as local to the given offset', function() {
      var inTokyo = new DateWithOffset(2008, 10, 22, 21, 0, 540);
      assert.equal(inTokyo.toISOString(), '2008-11-22T12:00:00.000Z');
      assert.equal(inTokyo.getHours(), 21);
    });
  });

});


test/daylight_savings_test.js
var assert = require("assert"),
    DateWithOffset = require('../');

describe('DateWithOffset, with an offset that changes over time', function() {

  "use strict";

  var zone = function(utcDate) {
    // DST April 1 through October 31
    var month = utcDate.getMonth();
    return (month < 3 || month > 9) ? 0 : +60;
  };

  it('adjusts the local time for daylight-savings', function() {
    var standard = new DateWithOffset(new Date("2011-02-12T18:00Z"), zone);
    assert.equal(standard.getTimezoneOffset(), 0);
    assert.equal(standard.toString(), 'Sat Feb 12 2011 18:00:00 GMT+0000');

    var daylight = new DateWithOffset(new Date("2011-07-12T18:00Z"), zone);
    assert.equal(daylight.getTimezoneOffset(), -60);
    assert.equal(daylight.toString(), 'Tue Jul 12 2011 19:00:00 GMT+0100');
  });

  it('adjusts when the date is moved', function() {
    var date = new DateWithOffset(new Date("2011-02-12T18:00Z"), zone);
    date.setMonth(6);
    assert.equal(date.getTimezoneOffset(), -60);
    assert.equal(date.toString(), 'Tue Jul 12 2011 19:00:00 GMT+0100');
  });

});


index.js
(function(window, undefined) {

  "use strict";

  var slice = Array.prototype.slice,
      MILLISECONDS_PER_MINUTE = 60 * 1000,
      OFFSET_SUFFIX = /(((GMT)?[\+\-]\d\d:?\d\d)|Z)(\s*\(.+\))?$/;

  function isNumber(x) { return typeof(x) === 'number'; }

  function isFunction(x) { return typeof(x) === 'function'; }

  function reifyOffset(date, offset) {
    return +(isFunction(offset) ? offset(date) : offset);
  }

  function isOffset(x) {
    if (x == null) { return false; }
    if (isFunction(x)) { return true; }
    if (isFunction(x.valueOf)) {
      x = x.valueOf();
    }
    return isNumber(x);
  }

  function applyOffset(date, offset) {
    offset = reifyOffset(date, offset);
    date.setTime( date.getTime() + MILLISECONDS_PER_MINUTE * offset );
    return date;
  }

  function buildDate(args, offset) {
    if (args.length === 0) { return new Date(); }

    if (args.length === 1 && args[0] instanceof Date) { return args[0]; }
    if (args.length === 1 && isNumber(args[0]))       { return new Date(args[0]); }

    if (args.length > 1) {
      args[3] = args[3] || null;
      args[4] = args[4] || null;
      args[5] = args[5] || null;
      args[6] = args[6] || null;

      date = new Date(args[0], args[1], args[2], args[3], args[4], args[5], args[6]);
      offset = reifyOffset(date, offset);
      return applyOffset( date, -date.getTimezoneOffset() - offset );
    }

    var string            = args[0].toString(),
        date              = new Date(string),
        isYYYYmmdd        = /\d\d\d\d-\d\d-\d\d/.test(string),
        isOffsetSpecified = OFFSET_SUFFIX.test(string),
        isLocal           = !isYYYYmmdd && !isOffsetSpecified;

    if (isLocal) {
      offset = reifyOffset(date, offset);
      date = applyOffset(date, -date.getTimezoneOffset() - offset);
    }

    return date;
  }

  function formattedOffset(offsetInMinutes) {
    var sign    = offsetInMinutes >= 0 ? '+' : '-';
    offsetInMinutes = Math.abs(offsetInMinutes);
    var hours   = Math.floor(offsetInMinutes / 60),
        minutes = offsetInMinutes - 60 * hours;
    if (hours < 10)   { hours = '0' + hours; }
    if (minutes < 10) { minutes = '0' + minutes; }
    return 'GMT' + sign + hours + minutes;
  }

  function DateWithOffset() {
    var args = slice.call(arguments, 0),
        offset = args.pop();

    if ( !isOffset(offset) ) {
      throw new TypeError('DateWithOffset requires an offset');
    }

    this.setTime(buildDate(args, offset));
    this.offset = function() { return offset; };
  }

  DateWithOffset.prototype = {

    // A Date whose UTC time is the local time of this object's real time.
    // That is, it is incorrect by `offset` minutes. Used for `getDate` et al.
    localDate: function() {
      var offset = reifyOffset(this.date(), this.offset());
      return applyOffset(this.date(), offset);
    },

    withOffset: function(offset) {
      return new DateWithOffset(this.getTime(), offset);
    },

    getTime:            function() { return this.date().getTime(); },
    toISOString:        function() { return this.date().toISOString(); },
    valueOf:            function() { return this.getTime(); },
    toJSON:             function() { return this.toISOString(); },

    getTimezoneOffset:  function() {
      return -reifyOffset(this, this.offset());
    },

    toString: function() {
      var localDate = this.localDate(),
          plusBrowserOffset = applyOffset(localDate, localDate.getTimezoneOffset()),
          asString = plusBrowserOffset.toString(),
          offset = reifyOffset(this, this.offset());
      return asString.replace(OFFSET_SUFFIX, formattedOffset(offset));
    },

    getYear: function() {
      return this.localDate().getUTCFullYear() - 1900;
    },

    setYear: function(year) {
      return this.setFullYear(1900 + year);
    },

    setTime: function(date) {
      this.date = function() { return new Date(date); };
      return this;
    }

  };

  function addGetters(property) {
    DateWithOffset.prototype['get' + property] = function() {
      return this.localDate()['getUTC' + property]();
    };

    DateWithOffset.prototype['getUTC' + property] = function() {
      return this.date()['getUTC' + property]();
    };
  }

  function addSetters(property) {
    DateWithOffset.prototype['set' + property] = function(newValue) {
      var localDate = this.localDate();
      localDate['setUTC' + property](newValue);
      var offset = reifyOffset(this, this.offset());
      return this.setTime( applyOffset(localDate, -offset) );
    };

    DateWithOffset.prototype['setUTC' + property] = function(newValue) {
      var date = this.date();
      date['setUTC' + property](newValue);
      return this.setTime(date);
    };
  }

  addGetters('Date');           addSetters('Date');
  addGetters('Day');            // can't set day of week
  addGetters('FullYear');       addSetters('FullYear');
  addGetters('Hours');          addSetters('Hours');
  addGetters('Milliseconds');   addSetters('Milliseconds');
  addGetters('Minutes');        addSetters('Minutes');
  addGetters('Month');          addSetters('Month');
  addGetters('Seconds');        addSetters('Seconds');

  if (typeof(module) !== 'undefined') {
    module.exports = DateWithOffset;
  }

  if (typeof(window) !== 'undefined') {
    window.DateWithOffset = DateWithOffset;
  }

}(this));
